### PR TITLE
Team 9 Jathurshan Towards Interpretable Sleep Stage Classification Using Cross-Modal Transformers

### PR DESCRIPTION
**Pull request by Jathurshan0330**

### CODE REVIEW

**Comment by Jathurshan0330** [comment]
Added  task function: multi_epoch_multi_modal_sleep_staging_sleepedf_fn, Cross-modal transformer, interactive plotting to visualize interpretations and multi_epoch_multi_modal_sleep_staging_sleepedf_fn.py

File: pyhealth/models/cross_modal_transformer.py [hunk 1]
```diff
@@ -0,0 +1,847 @@
...
+from torch import Tensor
+from torch.nn import functional as F
+from torch.nn import Module
+from torch.nn import MultiheadAttention
+from torch.nn import ModuleList
+from torch.nn.init import xavier_uniform_
+from torch.nn import Dropout
+from torch.nn import Linear
+from torch.nn import LayerNorm, BatchNorm1d
+from einops import rearrange, reduce, repeat
```
Comment (zlin7) [inline pyhealth/models/cross_modal_transformer.py]: Should add the dependency or pop up a message for the user to install this package. 

Is it possible to get rid of this dependency actually? For example, `Rearrange` can probably be achieved via `torch.permute`.

File: pyhealth/models/cross_modal_transformer.py [hunk 2]
```diff
@@ -0,0 +1,847 @@
...
+        Args:
+            kwargs: dict, input data
+        Returns:
+            results: dict, model outputs
+        '''
+        signal = torch.tensor(kwargs[self.feature_keys[0]],device=self.device)
+        y_true = torch.tensor(np.array(kwargs[self.label_key[0]]),device=self.device) # (B, num_epoch_seq)
+
+
+        # Get EEG and EOG signals and slice them into 30s epochs
```
Comment (zlin7) [inline pyhealth/models/cross_modal_transformer.py:530]: Can we apply this on any two types of signals? e.g. EEG + ECG? If so, maybe renaming the variables could be helpful.

File: pyhealth/tasks/sleep_staging.py [hunk 3]
```diff
@@ -358,6 +501,23 @@ def sleep_staging_shhs_fn(record, epoch_seconds=30):
     # # print(sleep_staging_ds.patient_to_index)
     # # print(sleep_staging_ds.record_to_index)
     # print(sleep_staging_ds.input_info)
+
```
Comment (zlin7) [inline pyhealth/tasks/sleep_staging.py]: Can remove this now

File: examples/sleep_staging_manytomany_sleepedf_cross_modal_transformer.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+from pyhealth.models.cross_modal_transformer import interpret_cmt, interactive_plot_cmt
+
+'''
+Example Colab tutorial can be found at:
+https://colab.research.google.com/drive/1jwD5NX8cR47MRtnOW_drKIS6lP1tbWFc?usp=sharing
+'''
+
+# step 1: load signal data
+sleepedf_ds = SleepEDFDataset(
+    root="/srv/local/data/SLEEPEDF/sleep-edf-database-expanded-1.0.0/sleep-cassette",#'/home/jp65/pyhealth_hackathon/storage.googleapis.com/pyhealth/sleepedf-sample',#"/srv/local/data/SLEEPEDF/sleep-edf-database-expanded-1.0.0/sleep-cassette",
```
Comment (zlin7) [inline examples/sleep_staging_manytomany_sleepedf_cross_modal_transformer.py]: let's clean up this path ( in particular the username)

File: pyhealth/models/cross_modal_transformer.py [hunk 5]
```diff
@@ -0,0 +1,847 @@
...
+        Returns:
+            Tensor, sliced signal
+        '''
+        sample_length = sample_rate * epoch_seconds
+        sliced_seq = []
+        for slice_index in range(signal.shape[-1] // sample_length):
+            sliced_seq.append(signal[:,:,slice_index * sample_length : (slice_index + 1) * sample_length])
+        return torch.cat(sliced_seq,dim = 1)
+    
+    def forward(self,**kwargs):
```
Comment (zlin7) [inline pyhealth/models/cross_modal_transformer.py:519]: Please also handle the kwarg `embed` - see https://github.com/sunlabuiuc/PyHealth/blob/89ee19fa084f4cc0be8205eb59b82375ff9e2c69/pyhealth/models/retain.py#L364. This is for the calibration module. Thanks!

File: pyhealth/models/cross_modal_transformer.py [hunk 6]
```diff
@@ -0,0 +1,847 @@
...
+from matplotlib.collections import LineCollection
+import matplotlib.pyplot as plt
+from IPython import display
+from ipywidgets import  interactive, IntSlider
+
+
+
+'''
+Colab tutorial can be found at:
+https://colab.research.google.com/drive/1jwD5NX8cR47MRtnOW_drKIS6lP1tbWFc?usp=sharing
```
Comment (zlin7) [inline pyhealth/models/cross_modal_transformer.py:30]: Please add more explanations to the tutorial

File: pyhealth/sleep_staging_manytomany_sleepedf_cross_modal_transformer.py
Comment (zlin7) [inline pyhealth/sleep_staging_manytomany_sleepedf_cross_modal_transformer.py:1]: Can we remove this file?

**Review by zlin7 — CHANGES_REQUESTED** [review]
Overall great job! Please take a look at my comments, and remove code that is commented out.

File: pyhealth/models/cross_modal_transformer.py (hunk 2 above)
Comment (Jathurshan0330) [reply pyhealth/models/cross_modal_transformer.py:530]: In the paper, it is mainly used for two signals: EEG and ECG. Due to that, I maintained the variables' names as it is also more easier to extract necessary attention weights for interpretation.

File: pyhealth/sleep_staging_manytomany_sleepedf_cross_modal_transformer.py
Comment (Jathurshan0330) [reply pyhealth/sleep_staging_manytomany_sleepedf_cross_modal_transformer.py:1]: It can be removed. Also, where can I mention about the colab tutorial

**Comment by ycq091044** [comment]
@Jathurshan0330 I will merge this PR to a temporary PR for local testing. Please follow Zhen's comments and finish the PR improvements if possible.

**Pull request merged**